File: robot_arm_node/commander/track_target_server.py
#!/usr/bin/env python3
"""
@file   track_target_server.py
@brief  ArmTrackTarget Action Server —— 视觉伺服目标跟随执行逻辑

职责：
  - 接收 goal 后通过 ROS2 参数服务启动 visp_ibvs_node（paused=False）
  - 订阅 feature 话题，10Hz 计算图像误差 / 深度误差，发送 Feedback
  - 监测收敛 / 特征丢失 / 超时 / cancel，退出时挂起 IBVS（paused=True）
  - 通过 StatusAggregator 将跟随状态写入 ArmStatus 广播

被 ArmCommanderNode 持有，不独立运行。

@version 1.0
@date 2026-06-23
@copyright Copyright (c) 2026 eMeet
"""

import math
import time
import threading

from rclpy.node import Node
from rcl_interfaces.srv import SetParameters
from rcl_interfaces.msg import Parameter, ParameterValue, ParameterType
from geometry_msgs.msg import PointStamped

from robot_arm_interfaces.action import ArmTrackTarget
from robot_arm_interfaces.msg import ArmStatus

from .status_aggregator import StatusAggregator


# ── 常量（与 visp_ibvs_node.cpp 保持一致）────────────────────────────────────
IMG_STOP_TH      = 0.005   # 图像误差收敛阈值（归一化）
DEPTH_STOP_TH    = 0.02    # 深度误差收敛阈值（m）
FEATURE_TIMEOUT  = 0.5     # 特征丢失判定时长（s）
FEEDBACK_HZ      = 10.0    # Feedback 发布频率
DEFAULT_DEPTH    = 0.3     # 未指定 desired_depth 时的默认值（m）

IBVS_NODE        = 'visp_ibvs_node'
FEATURE_TOPIC    = '/red_detector/feature'


class TrackTargetServer:
    """ArmTrackTarget Action 的执行逻辑。"""

    def __init__(self, node: Node, status: StatusAggregator):
        self._node   = node
        self._status = status
        self._logger = node.get_logger()

        # 特征缓存（线程安全）
        self._feat_lock      = threading.Lock()
        self._feat_x         = 0.0
        self._feat_y         = 0.0
        self._feat_z         = DEFAULT_DEPTH
        self._last_feat_time = None
        self._has_feat       = False

        # 取消标志（供 cancel 回调写入）
        self._cancel_flag = threading.Event()

        # 订阅特征话题
        node.create_subscription(
            PointStamped, FEATURE_TOPIC,
            self._on_feature, 10,
        )

        # visp_ibvs_node 参数服务客户端
        self._param_cli = node.create_client(
            SetParameters, f'/{IBVS_NODE}/set_parameters',
        )

    # ── 特征订阅回调 ──────────────────────────────────────────────────────────
    def _on_feature(self, msg: PointStamped):
        if not math.isfinite(msg.point.z) or msg.point.z <= 0.0:
            return
        with self._feat_lock:
            self._feat_x         = msg.point.x
            self._feat_y         = msg.point.y
            self._feat_z         = msg.point.z
            self._last_feat_time = time.time()
            self._has_feat       = True

    # ── IBVS 参数设置 ─────────────────────────────────────────────────────────
    def _set_params(self, **kwargs) -> bool:
        """向 visp_ibvs_node 写入 ROS2 参数。服务不可用时（仿真）直接返回 True。"""
        if not self._param_cli.wait_for_service(timeout_sec=1.0):
            self._logger.info(f'{IBVS_NODE} 参数服务不可用，仿真模式跳过')
            return True

        params = []
        for name, value in kwargs.items():
            p = Parameter(name=name)
            if isinstance(value, bool):
                p.value = ParameterValue(
                    type=ParameterType.PARAMETER_BOOL, bool_value=value)
            elif isinstance(value, float):
                p.value = ParameterValue(
                    type=ParameterType.PARAMETER_DOUBLE, double_value=value)
            params.append(p)

        req = SetParameters.Request(parameters=params)
        event  = threading.Event()
        result = [None]

        def _done(future):
            try:
                result[0] = future.result()
            except Exception:
                pass
            event.set()

        self._param_cli.call_async(req).add_done_callback(_done)
        if event.wait(timeout=2.0) and result[0]:
            return all(r.successful for r in result[0].results)
        return False

    # ── 取消接口（供 Commander cancel 回调调用）───────────────────────────────
    def cancel(self):
        self._cancel_flag.set()

    # ── 主执行入口 ────────────────────────────────────────────────────────────
    def execute(self, goal_handle) -> ArmTrackTarget.Result:
        """在 action 专用线程中阻塞执行，直到退出条件满足。"""
        goal = goal_handle.request
        self._cancel_flag.clear()

        result    = ArmTrackTarget.Result()
        img_err   = 999.0
        depth_err = 999.0

        # ── 配置并启动 IBVS ──────────────────────────────────────────────────
        desired_depth = float(goal.desired_depth) if goal.desired_depth > 0.0 \
                        else DEFAULT_DEPTH
        params = {
            'paused':     False,
            'desired_x':  float(goal.desired_x),
            'desired_y':  float(goal.desired_y),
            'desired_depth': desired_depth,
            'constrain_height': bool(goal.constrain_height),
        }
        if goal.constrain_height and goal.desired_height > 0.0:
            params['desired_height'] = float(goal.desired_height)

        if not self._set_params(**params):
            self._logger.error('IBVS 参数设置失败，无法启动跟随')
            result.success    = False
            result.exit_code  = ArmTrackTarget.Result.EXIT_ERROR
            result.exit_reason = 'error'
            goal_handle.abort()
            return result

        self._status.set_tracking(True, 0.0, 0.0)
        self._logger.info(
            f'目标跟随启动  depth={desired_depth:.2f}m  '
            f'hold={goal.hold_on_converge}  timeout={goal.total_timeout_sec:.1f}s')

        # ── 主控制循环 ───────────────────────────────────────────────────────
        t_start   = time.time()
        fb_period = 1.0 / FEEDBACK_HZ
        last_fb   = 0.0

        while True:
            now     = time.time()
            elapsed = now - t_start

            # 急停检测（commander 进入 STOPPED）：立即退出，清理段会 paused=True 停 IBVS
            if self._node.is_stopped():
                self._logger.info('目标跟随期间被急停，退出')
                result.exit_code   = ArmTrackTarget.Result.EXIT_ERROR
                result.exit_reason = 'stopped'
                result.success     = False
                break

            # 取消检测
            if goal_handle.is_cancel_requested or self._cancel_flag.is_set():
                self._logger.info('目标跟随被取消')
                result.exit_code   = ArmTrackTarget.Result.EXIT_CANCELLED
                result.exit_reason = 'cancelled'
                result.success     = True
                break

            # 总超时
            if goal.total_timeout_sec > 0.0 and elapsed >= goal.total_timeout_sec:
                self._logger.warn(f'目标跟随总超时（{goal.total_timeout_sec:.1f}s）')
                result.exit_code   = ArmTrackTarget.Result.EXIT_TIMEOUT
                result.exit_reason = 'timeout'
                result.success     = False
                break

            # 读取特征快照
            with self._feat_lock:
                has_feat  = self._has_feat
                last_t    = self._last_feat_time
                fx, fy, fz = self._feat_x, self._feat_y, self._feat_z

            # 特征丢失
            if has_feat and (now - last_t) > FEATURE_TIMEOUT:
                self._logger.warn('特征丢失超时，退出跟随')
                result.exit_code   = ArmTrackTarget.Result.EXIT_FEATURE_LOST
                result.exit_reason = 'feature_lost'
                result.success     = False
                break

            # 误差计算
            if has_feat:
                img_err   = math.hypot(fx - goal.desired_x, fy - goal.desired_y)
                depth_err = abs(math.log(fz / desired_depth)) if fz > 0 else 999.0
            else:
                img_err = depth_err = 999.0

            self._status.set_tracking(True, img_err, depth_err)

            # 收敛检测
            is_converged = img_err < IMG_STOP_TH and depth_err < DEPTH_STOP_TH
            if is_converged and not goal.hold_on_converge:
                self._logger.info(
                    f'目标收敛  img_err={img_err:.4f}  depth_err={depth_err:.3f}m')
                result.exit_code   = ArmTrackTarget.Result.EXIT_CONVERGED
                result.exit_reason = 'converged'
                result.success     = True
                break

            # Feedback
            if now - last_fb >= fb_period:
                fb = ArmTrackTarget.Feedback()
                fb.img_err      = float(img_err)
                fb.depth_err_m  = float(depth_err)
                fb.elapsed_sec  = float(elapsed)
                fb.is_converged = bool(is_converged)
                fb.current_pose = self._status.pose
                goal_handle.publish_feedback(fb)
                last_fb = now

            time.sleep(0.02)

        # ── 退出清理 ─────────────────────────────────────────────────────────
        self._set_params(paused=True)
        self._status.set_tracking(False, 0.0, 0.0)

        result.final_img_err     = float(img_err)
        result.final_depth_err_m = float(depth_err)
        result.final_pose        = self._status.pose

        if result.exit_code == ArmTrackTarget.Result.EXIT_CANCELLED:
            goal_handle.canceled()
        elif result.success:
            goal_handle.succeed()
        else:
            goal_handle.abort()

        self._logger.info(
            f'目标跟随结束  reason={result.exit_reason}  '
            f'img_err={result.final_img_err:.4f}')
        return result

File: robot_arm_node/src/commander/track_target_server.cpp
/**
 * @file track_target_server.cpp
 * @brief TrackTargetServer 实现 —— IBVS 目标跟随控制循环
 *
 * SetParameters 启动 IBVS（paused=false）→ 主循环 10Hz：急停/取消/超时/特征丢失检查，
 * 图像+深度误差计算与收敛判定，发 Feedback；退出时 paused=true 挂起 IBVS 并调 goal 终态。
 *
 * @version 1.0
 * @date 2026-07-01
 * @copyright Copyright (c) 2026 eMeet
 */
#include "robot_arm_node/commander/track_target_server.hpp"

#include <chrono>
#include <cmath>
#include <thread>

#include <rcl_interfaces/msg/parameter.hpp>
#include <rcl_interfaces/msg/parameter_type.hpp>

namespace robot_arm_node::commander
{

using namespace std::chrono_literals;
namespace
{
constexpr double IMG_STOP_TH     = 0.005;   // 图像误差收敛阈值（归一化）
constexpr double DEPTH_STOP_TH   = 0.02;    // 深度误差收敛阈值（m）
constexpr double FEATURE_TIMEOUT = 0.5;     // 特征丢失判定时长（s）
constexpr double FEEDBACK_HZ     = 10.0;
constexpr double DEFAULT_DEPTH   = 0.3;     // 未指定 desired_depth 时的默认值（m）
const char * IBVS_NODE     = "visp_ibvs_node";
const char * FEATURE_TOPIC = "/red_detector/feature";

/**
 * @brief 取单调时钟的当前秒数。
 *
 * 用 steady_clock 而非 ROS 时钟：本循环是 wall timer 驱动，且仿真下
 * /clock 频率很低，用 ROS 时钟求时间差会系统性丢步（见 CLAUDE.md）。
 *
 * @return 自 steady_clock 纪元起的秒数。
 */
double now_s()
{
  return std::chrono::duration<double>(std::chrono::steady_clock::now().time_since_epoch()).count();
}

/**
 * @brief 构造一个 bool 类型的 ROS 参数消息。
 *
 * @param name 参数名。
 * @param v 参数值。
 * @return 填好类型与值的 Parameter 消息。
 */
rcl_interfaces::msg::Parameter bool_param(const std::string & name, bool v)
{
  rcl_interfaces::msg::Parameter p;
  p.name = name;
  p.value.type = rcl_interfaces::msg::ParameterType::PARAMETER_BOOL;
  p.value.bool_value = v;
  return p;
}
/**
 * @brief 构造一个 double 类型的 ROS 参数消息。
 *
 * @param name 参数名。
 * @param v 参数值。
 * @return 填好类型与值的 Parameter 消息。
 */
rcl_interfaces::msg::Parameter double_param(const std::string & name, double v)
{
  rcl_interfaces::msg::Parameter p;
  p.name = name;
  p.value.type = rcl_interfaces::msg::ParameterType::PARAMETER_DOUBLE;
  p.value.double_value = v;
  return p;
}
}  // namespace

/**
 * @brief 构造目标跟随 Action 服务端，订阅特征话题并建立 IBVS 参数客户端。
 *
 * 深度初值给 DEFAULT_DEPTH 而非 0：0 会让首帧的深度误差算出 log(0) = -inf。
 *
 * @param node 宿主节点。
 * @param status 状态聚合器，用于上报跟随状态与读当前位姿。
 * @param is_stopped 急停判据回调。
 */
TrackTargetServer::TrackTargetServer(rclcpp::Node & node, state::StatusAggregator & status,
                                     std::function<bool()> is_stopped)
: node_(node), logger_(node.get_logger()), status_(status), is_stopped_(std::move(is_stopped))
{
  feat_z_ = DEFAULT_DEPTH;
  feat_sub_ = node_.create_subscription<geometry_msgs::msg::PointStamped>(
      FEATURE_TOPIC, 10,
      [this](const geometry_msgs::msg::PointStamped & msg) { this->on_feature(msg); });
  param_cli_ = node_.create_client<SetParameters>(std::string("/") + IBVS_NODE + "/set_parameters");
}

/**
 * @brief 特征话题回调：缓存目标在图像上的归一化坐标与深度。
 *
 * 丢弃非有限值与非正深度 —— 检测器在目标丢失时可能发 NaN/0，
 * 混进缓存会让误差计算和收敛判定失效。
 *
 * @param msg 特征点消息（x/y 为归一化图像坐标，z 为深度 m）。
 */
void TrackTargetServer::on_feature(const geometry_msgs::msg::PointStamped & msg)
{
  if (!std::isfinite(msg.point.z) || msg.point.z <= 0.0) return;
  std::lock_guard<std::mutex> lk(feat_mtx_);
  feat_x_ = msg.point.x; feat_y_ = msg.point.y; feat_z_ = msg.point.z;
  last_feat_s_ = now_s();
  has_feat_ = true;
}

/**
 * @brief 批量设置 IBVS 节点的参数。
 *
 * 参数服务不可用时返回 true（视为成功）—— 仿真或未起 visp_ibvs_node 的场景
 * 下不应因此让整个动作失败，只打一条 INFO。
 *
 * @param params 待设置的参数列表。
 * @return 全部设置成功（或服务不存在）返回 true，任一失败/超时返回 false。
 */
bool TrackTargetServer::set_params(const std::vector<rcl_interfaces::msg::Parameter> & params)
{
  if (!param_cli_->wait_for_service(1s)) {
    RCLCPP_INFO(logger_, "%s 参数服务不可用，仿真模式跳过", IBVS_NODE);
    return true;
  }
  auto req = std::make_shared<SetParameters::Request>();
  req->parameters = params;
  auto future = param_cli_->async_send_request(req);
  if (future.wait_for(2s) != std::future_status::ready) {
    param_cli_->remove_pending_request(future);
    return false;
  }
  auto resp = future.get();
  if (!resp) return false;
  for (const auto & r : resp->results) {
    if (!r.successful) return false;
  }
  return true;
}

/**
 * @brief 挂起或恢复 IBVS 视觉伺服。
 *
 * @param paused true 挂起（停止发速度指令），false 恢复。
 * @return 设置成功返回 true。
 */
bool TrackTargetServer::set_paused(bool paused)
{
  return set_params({bool_param("paused", paused)});
}

/**
 * @brief 请求中止当前跟随（线程安全，供外部调用）。
 *
 * 只置标志位，由主循环下一拍检测并退出，不直接操作 Action 句柄。
 */
void TrackTargetServer::cancel()
{
  cancel_flag_.store(true);
}

/**
 * @brief 执行 TrackTarget 动作：启动 IBVS → 10Hz 监控循环 → 退出时挂起 IBVS。
 *
 * 本服务端自己不算控制量，速度指令由 visp_ibvs_node 直接发出；这里只负责
 * 启停 IBVS、监控误差与各种退出条件、上报 Feedback。循环内检查顺序为
 * 急停 → 取消 → 总超时 → 特征丢失 → 误差与收敛 → Feedback。
 *
 * hold_on_converge 为 true 时收敛也不退出（持续保持对准），此时只能靠取消、
 * 超时、特征丢失或急停结束动作。无论从哪条路径退出，都必须走到函数末尾的
 * 清理段把 IBVS 挂起 —— 否则视觉伺服会在动作结束后继续驱动机械臂。
 *
 * @param gh Action 目标句柄。
 * @return Action 结果（含 exit_code、最终误差与位姿）。
 */
TrackTargetServer::Action::Result TrackTargetServer::execute(const std::shared_ptr<GoalHandle> & gh)
{
  const auto goal = gh->get_goal();
  cancel_flag_.store(false);

  Action::Result result;
  double img_err = 999.0, depth_err = 999.0;

  // 配置并启动 IBVS
  const double desired_depth = goal->desired_depth > 0.0 ? goal->desired_depth : DEFAULT_DEPTH;
  std::vector<rcl_interfaces::msg::Parameter> params{
      bool_param("paused", false),
      double_param("desired_x", goal->desired_x),
      double_param("desired_y", goal->desired_y),
      double_param("desired_depth", desired_depth),
      bool_param("constrain_height", goal->constrain_height)};
  if (goal->constrain_height && goal->desired_height > 0.0) {
    params.push_back(double_param("desired_height", goal->desired_height));
  }

  if (!set_params(params)) {
    RCLCPP_ERROR(logger_, "IBVS 参数设置失败，无法启动跟随");
    result.success = false;
    result.exit_code = Action::Goal::EXIT_ERROR;
    result.exit_reason = "error";
    gh->abort(std::make_shared<Action::Result>(result));
    return result;
  }

  status_.set_tracking(true, 0.0, 0.0);
  RCLCPP_INFO(logger_, "目标跟随启动  depth=%.2fm  hold=%d  timeout=%.1fs",
              desired_depth, goal->hold_on_converge, goal->total_timeout_sec);

  // 主控制循环
  const double t_start = now_s();
  const double fb_period = 1.0 / FEEDBACK_HZ;
  double last_fb = 0.0;

  while (true) {
    const double now = now_s();
    const double elapsed = now - t_start;

    // 急停
    if (is_stopped_ && is_stopped_()) {
      RCLCPP_INFO(logger_, "目标跟随期间被急停，退出");
      result.exit_code = Action::Goal::EXIT_ERROR; result.exit_reason = "stopped";
      result.success = false; break;
    }
    // 取消
    if (gh->is_canceling() || cancel_flag_.load()) {
      RCLCPP_INFO(logger_, "目标跟随被取消");
      result.exit_code = Action::Goal::EXIT_CANCELLED; result.exit_reason = "cancelled";
      result.success = true; break;
    }
    // 总超时
    if (goal->total_timeout_sec > 0.0 && elapsed >= goal->total_timeout_sec) {
      RCLCPP_WARN(logger_, "目标跟随总超时（%.1fs）", goal->total_timeout_sec);
      result.exit_code = Action::Goal::EXIT_TIMEOUT; result.exit_reason = "timeout";
      result.success = false; break;
    }

    // 特征快照
    bool has_feat; double last_t, fx, fy, fz;
    {
      std::lock_guard<std::mutex> lk(feat_mtx_);
      has_feat = has_feat_; last_t = last_feat_s_;
      fx = feat_x_; fy = feat_y_; fz = feat_z_;
    }

    // 特征丢失
    if (has_feat && (now - last_t) > FEATURE_TIMEOUT) {
      RCLCPP_WARN(logger_, "特征丢失超时，退出跟随");
      result.exit_code = Action::Goal::EXIT_FEATURE_LOST; result.exit_reason = "feature_lost";
      result.success = false; break;
    }

    // 误差计算
    if (has_feat) {
      img_err   = std::hypot(fx - goal->desired_x, fy - goal->desired_y);
      depth_err = fz > 0.0 ? std::fabs(std::log(fz / desired_depth)) : 999.0;
    } else {
      img_err = depth_err = 999.0;
    }
    status_.set_tracking(true, img_err, depth_err);

    // 收敛检测
    const bool is_converged = img_err < IMG_STOP_TH && depth_err < DEPTH_STOP_TH;
    if (is_converged && !goal->hold_on_converge) {
      RCLCPP_INFO(logger_, "目标收敛  img_err=%.4f  depth_err=%.3fm", img_err, depth_err);
      result.exit_code = Action::Goal::EXIT_CONVERGED; result.exit_reason = "converged";
      result.success = true; break;
    }

    // Feedback
    if (now - last_fb >= fb_period) {
      auto fb = std::make_shared<Action::Feedback>();
      fb->img_err      = static_cast<float>(img_err);
      fb->depth_err_m  = static_cast<float>(depth_err);
      fb->elapsed_sec  = static_cast<float>(elapsed);
      fb->is_converged = is_converged;
      fb->current_pose = status_.pose();
      gh->publish_feedback(fb);
      last_fb = now;
    }

    std::this_thread::sleep_for(20ms);
  }

  // 退出清理：无论何种退出路径都必须挂起 IBVS，否则它会继续驱动机械臂
  // 退出清理
  set_paused(true);
  status_.set_tracking(false, 0.0, 0.0);

  result.final_img_err     = static_cast<float>(img_err);
  result.final_depth_err_m = static_cast<float>(depth_err);
  result.final_pose        = status_.pose();

  // 按退出原因调对应的终态方法：取消走 canceled，其余按成败 succeed/abort
  auto result_ptr = std::make_shared<Action::Result>(result);
  if (result.exit_code == Action::Goal::EXIT_CANCELLED) {
    gh->canceled(result_ptr);
  } else if (result.success) {
    gh->succeed(result_ptr);
  } else {
    gh->abort(result_ptr);
  }

  RCLCPP_INFO(logger_, "目标跟随结束  reason=%s  img_err=%.4f",
              result.exit_reason.c_str(), result.final_img_err);
  return result;
}

}  // namespace robot_arm_node::commander
